sphinx_immaterial/custom_admonitions.py

```diff
@@ -0,0 +1,354 @@
+    if getattr(config, "sphinx_immaterial_generate_inherited_admonitions"):
+        for admonition in (
+            "abstract",
+            "info",
+            "success",
+            "question",
+            "failure",
+            "bug",
+            "example",
+            "quote",
+        ):
+            app.add_directive(
+                f"si-{admonition}",
+                get_directive_class(admonition, _(admonition.title())),
+            )
+
+    user_defined_admonitions = getattr(config, "sphinx_immaterial_custom_admonitions")
+    user_defined_dir_names = [
+        directive["name"] for directive in user_defined_admonitions
+    ]
+
+    # override the specific admonitions defined in sphinx and docutils
+    # these are the admonitions that have translated titles in sphinx.locale
+    for admonition, title in admonitionlabels.items():
+
+        if admonition in user_defined_dir_names:
+            continue
+        app.add_directive(admonition, get_directive_class(admonition, title), True)
+
+
+def consolidate_css(app: Sphinx, env: BuildEnvironment):
+    """Generates the CSS for icons and admonitions, then appends that to the
+    theme's bundled CSS."""
+
+    theme_options = app.config["html_theme_options"]
+    is_palette_defined = False
+    if theme_options:
+        is_palette_defined = "palette" in theme_options
+
+    custom_admonitions = getattr(env, "sphinx_immaterial_custom_admonitions")
+    custom_icons = getattr(env, "sphinx_immaterial_custom_icons")
+    jinja_env = jinja2.Environment(
+        loader=jinja2.FileSystemLoader(str(PurePath(__file__).parent))
+    )
+    template = jinja_env.get_template("custom_admonitions.css")
+    generated = template.render(
+        icons=custom_icons,
+        admonitions=custom_admonitions,
+    )
+
+    css_name = "stylesheets/sphinx_immaterial_theme.min.css"
+    css_output = Path(app.outdir, "_static", css_name)
+    css_output.parent.mkdir(parents=True, exist_ok=True)
+    with open(str(css_output), mode="wb") as consolidated_out:
+        # copy pre-minified CSS from theme's bundles
+        theme_bundles = Path(__file__).parent / "static" / "stylesheets"
+        for bundle in theme_bundles.glob("*.min.css"):
+            if not is_palette_defined and bundle.name.startswith("palette"):
+                continue  # don't need the palette CSS if HTML doesn't need it
+            consolidated_out.write(bundle.read_bytes())
+
+        # append the generated CSS for icons and admonitions
+        consolidated_out.write(generated.replace("\n", "").encode(encoding="utf-8"))
+
+    # ensure new CSS file is added in the rendered HTML output
+    app.add_css_file(css_name)
+
+
+def setup(app: Sphinx):
+    """register our custom directive."""
+    app.add_config_value(
+        name="sphinx_immaterial_custom_admonitions",
+        default=[],
+        rebuild="env",
+        types=[List[Dict[str, Union[str, Tuple[int, int, int], bool]]]],
```

Comment:
Sphinx doesn't actually do any deep validation of the types --- so you can just specify `List[CustomAdmonition]` as the type and then our `confval` directive will show more useful type information.

Reply:
Every time I tried this, I get an error about CustomAdmonition class is not generic. It seems that the param needs to be a literal list of types. So, `types=[List[CustomAdmonitionConfig]],` will work, but it may be confusing because the config is actually a list of dicts, not a list of instantiated objects. 
![image](https://user-images.githubusercontent.com/14963867/201434132-10f23c0c-97e3-446f-aac3-41483c718e4d.png)